pytorch/pytorch
# [c2] framework for committed serialized tests

Summary:
Generate serialized test inputs/outputs/backward graphs of tests inside `caffe2/python/operator_test` that call assertSerializedOperatorCheck(). Tests should be decorated with serialized_test.collect_tests.given_and_seeded to run hypothesis tests that are actually random and a single fixed seeded hypothesis tests.

To use:
1. Refactor your test to be a SerializedTestCase
1a. Decorate it with @given_and_seeded
1b. Call testWithArgs in main
2. Run your test with -g to generate the output. Check it in.
3. Subsequent runs of the test without generating the output will check against the checked in test case. 

Details:
Run your test with `python caffe2/python/operator_test/[your_test].py -g`
Outputs are in `caffe2/python/serialized_test/data`. The operator tests outputs are in a further subdirectory `operator_test`, to allow for other tests in the future (model zoo tests?)

Currently, we've only refactored weighted_sum_test to use this, but in the next diff, we'll refactor as many as possible. The directory structure may also change as usually there are multiple tests in a single file, so we may create more structure to account for that.

Test Plan:
1. Run `python caffe2/python/operator_test/weighted_sum_test.py -g` to generate the outputs
2. Run `python caffe2/python/operator_test/weighted_sum_test.py`; Check that the unit test itself still runs as expected
3. Run `python caffe2/python/gradient_check_test.py`; Check that CheckSimple still works as expected
4. Run `with-proxy FULL_CAFFE2=1 python setup.py develop`; Check there are no errors
